packages/context/test/acceptance/class-level-bindings.ts

```diff
@@ -177,6 +184,216 @@ describe('Context bindings - Injecting dependencies of classes', () => {
+    class Store {
+      constructor(@inject.options('x') public optionX: number) {}
+    }
+
+    ctx
+      .bind('store')
+      .toClass(Store)
+      .withOptions({x: Promise.reject('invalid')});
+
+    try {
+      await ctx.get('store');
+      throw new Error('the test should fail');
+    } catch (e) {
+      expect(e).to.eql('invalid');
+    }
+  });
+
+  it('injects an option when `options` is a promise', async () => {
+    class Store {
+      constructor(@inject.options('x') public optionX: number) {}
+    }
+
+    ctx
+      .bind('store')
+      .toClass(Store)
+      .withOptions(Promise.resolve({x: 1}));
+    const store = await ctx.get('store');
+    expect(store.optionX).to.eql(1);
+  });
+
+  it('injects an option when `options` is a promise', async () => {
+    class Store {
+      constructor(@inject.options('x#y') public optionY: string) {}
+    }
+
+    ctx
+      .bind('store')
+      .toClass(Store)
+      .withOptions(Promise.resolve({x: {y: 'y'}}));
+    const store = await ctx.get('store');
+    expect(store.optionY).to.eql('y');
+  });
+
+  it('injects an option with nested property', () => {
+    class Store {
+      constructor(@inject.options('x#y') public optionXY: string) {}
+    }
+
+    ctx
+      .bind('store')
+      .toClass(Store)
+      .withOptions({x: {y: 'y'}});
+    const store = ctx.getSync('store');
+    expect(store.optionXY).to.eql('y');
+  });
+
+  it('injects options if the binding key is not present', () => {
+    class Store {
+      constructor(@inject.options() public options: object) {}
+    }
+
+    ctx
+      .bind('store')
+      .toClass(Store)
+      .withOptions({x: 1, y: 'a'});
+    const store = ctx.getSync('store');
+    expect(store.options).to.eql({x: 1, y: 'a'});
+  });
+
+  it("injects options if the binding key is ''", () => {
+    class Store {
+      constructor(@inject.options('') public options: object) {}
+    }
+
+    ctx
+      .bind('store')
+      .toClass(Store)
+      .withOptions({x: 1, y: 'a'});
+    const store = ctx.getSync('store');
+    expect(store.options).to.eql({x: 1, y: 'a'});
+  });
+
+  it('injects options if the binding key is a path', () => {
+    class Store {
+      constructor(@inject.options('#x') public optionX: number) {}
+    }
+
+    ctx
+      .bind('store')
+      .toClass(Store)
+      .withOptions({x: 1, y: 'a'});
+    const store = ctx.getSync('store');
+    expect(store.optionX).to.eql(1);
+  });
+
+  it('injects undefined option if key not found', () => {
+    class Store {
+      constructor(
+        // tslint:disable-next-line:no-any
+        @inject.options('not-exist') public option: string | undefined,
+      ) {}
+    }
+
+    ctx
+      .bind('store')
+      .toClass(Store)
+      .withOptions({x: 1, y: 'a'});
+    const store = ctx.getSync('store');
+    expect(store.option).to.be.undefined();
+  });
+
+  it('injects an option based on the parent binding', async () => {
+    class Store {
```

Comment:
hmm, I am confused how it tests "based on the parent binding", it looks like the same test case as https://github.com/strongloop/loopback-next/pull/671/files#diff-de8448577089742287c2819e357fb2eeR187

Reply:
In this instance, I think "parent" is a reference to the binding that holds whatever metadata is required to satisfy the different Store instances' injection targets with unique values.

I'd probably rename the test to be `injects options based on corresponding bindings` or something to that effect.